File: src/mapfile.cc
#include "mold.h"

#include <fstream>
#include <iomanip>
#include <ios>
#include <sstream>
#include <tbb/concurrent_hash_map.h>
#include <tbb/parallel_for_each.h>

namespace mold {

template <typename E>
using Map =
  tbb::concurrent_hash_map<InputSection<E> *, std::vector<Symbol<E> *>>;

template <typename E>
static Map<E> get_map(Context<E> &ctx) {
  Map<E> map;

  tbb::parallel_for_each(ctx.objs, [&](ObjectFile<E> *file) {
    for (Symbol<E> *sym : file->symbols) {
      if (sym->file != file || sym->get_type() == STT_SECTION)
        continue;

      if (InputSection<E> *isec = sym->get_input_section()) {
        assert(file == &isec->file);
        typename Map<E>::accessor acc;
        map.insert(acc, {isec, {}});
        acc->second.push_back(sym);
      }
    }
  });

  if (map.size() <= 1)
    return map;

  tbb::parallel_for(map.range(), [](const typename Map<E>::range_type &range) {
    for (auto &[k, v] : range)
      ranges::stable_sort(v, {}, &Symbol<E>::value);
  });
  return map;
}

template <typename E>
void print_map(Context<E> &ctx) {
  Timer t(ctx, "print_map");

  std::ostream *out = &std::cout;
  std::ofstream file;

  if (!ctx.arg.Map.empty()) {
    file.open(ctx.arg.Map);
    if (!file.is_open())
      Fatal(ctx) << "cannot open " << ctx.arg.Map << ": " << errno_string();
    out = &file;
  }

  // Construct a section-to-symbol map.
  Map<E> map = get_map(ctx);

  // Print a mapfile.
  *out << "               VMA       Size Align Out     In      Symbol\n";

  for (Chunk<E> *chunk : ctx.chunks) {
    *out << std::showbase
         << std::setw(18) << std::hex << (u64)chunk->shdr.sh_addr << std::dec
         << std::setw(11) << (u64)chunk->shdr.sh_size
         << std::setw(6) << (u64)chunk->shdr.sh_addralign
         << " " << chunk->name << "\n";

    OutputSection<E> *osec = chunk->to_osec();
    if (!osec)
      continue;

    std::span<InputSection<E> *> members = osec->members;
    std::vector<std::string> bufs(members.size());

    tbb::parallel_for((i64)0, (i64)members.size(), [&](i64 i) {
      InputSection<E> *mem = members[i];
      std::ostringstream ss;

      u64 addr = 0;
      if (osec->shdr.sh_flags & SHF_ALLOC)
        addr = osec->shdr.sh_addr + mem->offset;

      ss << std::showbase
         << std::setw(18) << std::hex << addr << std::dec
         << std::setw(11) << (u64)mem->sh_size
         << std::setw(6) << (1 << mem->p2align)
         << "         " << *mem << "\n";

      typename Map<E>::const_accessor acc;
      if (map.find(acc, mem))
        for (Symbol<E> *sym : acc->second)
          ss << std::showbase
             << std::setw(18) << std::hex << sym->get_addr(ctx) << std::dec
             << "          0     0                 "
             << *sym << "\n";

      bufs[i] = ss.str();
    });

    for (std::string &str : bufs)
      *out << str;
  }
}

using E = MOLD_TARGET;

template void print_map(Context<E> &ctx);

} // namespace mold

File: src/mapfile.rs
//! `--print-map` / `-Map`: a listing of output sections, their input
//! sections and symbols.

use std::collections::HashMap;
use std::fmt::Write;

use rayon::prelude::*;

use crate::arch::Arch;
use crate::chunks::ChunkId;
use crate::cmdline::ReportOutput;
use crate::context::Context;
use crate::elf::*;
use crate::input_files::FileId;
use crate::input_sections::InputSectionId;
use crate::symbol::SymbolId;

// Construct a section-to-symbol map.
fn section_symbols<E: Arch>(ctx: &Context<E>) -> HashMap<InputSectionId, Vec<SymbolId>> {
    let mut map: HashMap<InputSectionId, Vec<SymbolId>> = HashMap::new();
    for file in &ctx.objs {
        let file_id = FileId::Obj(file.id());
        for &id in &file.base.symbols {
            let sym = &ctx.symbols[id];
            if sym.file() == Some(file_id)
                && sym.ty() != STT_SECTION
                && let Some(r) = sym.input_section()
            {
                map.entry(r).or_default().push(id);
            }
        }
    }
    for syms in map.values_mut() {
        syms.sort_by_key(|&id| ctx.symbols[id].value);
    }
    map
}

pub fn print_map<E: Arch>(ctx: &Context<E>, output: &ReportOutput) {
    // Print a mapfile.
    let _t = ctx.timer("print_map");
    let map = section_symbols(ctx);

    output.with_writer("--print-map", |out| {
        out.write_all(b"               VMA       Size Align Out     In      Symbol\n")?;
        for &id in &ctx.chunks {
            let hdr = ctx.chunk_header(id);
            writeln!(
                out,
                "{:#18x}{:>11}{:>6} {}",
                hdr.shdr.sh_addr.get(),
                hdr.shdr.sh_size.get(),
                hdr.shdr.sh_addralign.get(),
                hdr.name
            )?;

            let ChunkId::Output(osec_id) = id else {
                continue;
            };
            let osec = &ctx.output_sections[osec_id.index()];
            let lines: Vec<String> = osec
                .members
                .par_iter()
                .map(|&member| {
                    let isec = ctx.input_section(member);
                    let addr = if osec.hdr.is_alloc() {
                        osec.hdr.shdr.sh_addr.get() + isec.offset()
                    } else {
                        0
                    };
                    let mut s = format!(
                        "{addr:#18x}{:>11}{:>6}         {}\n",
                        isec.sh_size,
                        1u64 << isec.p2align(),
                        isec.display(&ctx.objs[isec.file.index()])
                    );
                    if let Some(syms) = map.get(&member) {
                        for &id in syms {
                            let sym = &ctx.symbols[id];
                            writeln!(
                                s,
                                "{:#18x}          0     0                 {sym}",
                                sym.addr(ctx)
                            )
                            .unwrap();
                        }
                    }
                    s
                })
                .collect();
            for line in lines {
                out.write_all(line.as_bytes())?;
            }
        }
        Ok(())
    });
}
